cloud-nio/cloud-nio-impl-ftp/src/main/scala/cloud/nio/impl/ftp/FtpCloudNioFileProvider.scala

```diff
@@ -1,51 +1,53 @@
-    def ensureSlashedPrefix = if (cloudPath.startsWith("/")) cloudPath else s"/$cloudPath"
-    def ensureSlashed = if (cloudPath.endsWith("/")) cloudPath else s"$cloudPath/"
-  }
-}
-
 class FtpCloudNioFileProvider(fsProvider: FtpCloudNioFileSystemProvider) extends CloudNioFileProvider {
-  override def existsPath(cloudHost: String, cloudPath: String) = {
-    run(FtpOperation(cloudHost, cloudPath, "determine file existence"), acquireLease(cloudHost))(_.listFiles(cloudPath.ensureSlashedPrefix)).nonEmpty
-  }
-
-  override def existsPaths(cloudHost: String, cloudPathPrefix: String) = {
-    // We need to list the directories in the parent and see if any matches the name, hence the string manipulations
-    val parts = cloudPathPrefix.ensureSlashedPrefix.stripSuffix(CloudNioFileSystem.Separator).split(CloudNioFileSystem.Separator)
-    val parent = parts.init.mkString(CloudNioFileSystem.Separator)
-    val directoryName = parts.last
-
-    run(FtpOperation(cloudHost, cloudPathPrefix, "determine directory existence"), acquireLease(cloudHost))(_.listDirectories(parent))
-      .exists(_.getName == directoryName)
+  override def existsPath(cloudHost: String, cloudPath: String) = withAutoRelease(cloudHost) { client =>
+    FtpListFiles(cloudHost, cloudPath, "determine file existence")
+      .run(client)
+      .map(_.nonEmpty)
+  } unsafeRunSync()
+
+  override def existsPaths(cloudHost: String, cloudPathPrefix: String) = withAutoRelease(cloudHost) { client =>
+    existsPathsWithClient(cloudHost, cloudPathPrefix, client)
+  } unsafeRunSync()
+
+  private def existsPathsWithClient(cloudHost: String, cloudPathPrefix: String, client: FTPClient): IO[Boolean] = {
+    val operation = FtpListDirectories(cloudHost, cloudPathPrefix, "determine directory existence")
+    operation
+      .run(client)
+      .map(_.exists(_.getName == operation.directoryName))
   }
 
   /**
     * Returns a listing of keys within a bucket starting with prefix. The returned keys should include the prefix. The
     * paths must be absolute, but the key should not begin with a slash.
     */
-  override def listObjects(cloudHost: String, cloudPathPrefix: String, markerOption: Option[String]): CloudNioFileList = {
-    val files = run(FtpOperation(cloudHost, cloudPathPrefix, "list objects"), acquireLease(cloudHost)) { client =>
-      client.listFiles(cloudPathPrefix.ensureSlashedPrefix)
-    }
-
-    CloudNioFileList(files.map(_.getName).map(cloudPathPrefix.stripPrefix("/").ensureSlashed + _), markerOption)
-  }
+  override def listObjects(cloudHost: String, cloudPathPrefix: String, markerOption: Option[String]): CloudNioFileList = withAutoRelease(cloudHost) { client =>
+    FtpListFiles(cloudHost, cloudPathPrefix, "list objects")
+      .run(client)
+      .map(
+        _.map(_.getName).map(cloudPathPrefix.stripPrefix("/").ensureSlashed + _)
+      )
+      .map(
+        CloudNioFileList(_, markerOption)
+      )
```

Comment:
TOL That's a lot of `map`. Can some of these be combined?

Reply:
I'll take a look